python/pyspark/serializers.py

```diff
@@ -352,28 +364,17 @@ def dump_stream(self, iterator, stream):
         Make ArrowRecordBatches from Pandas Series and serialize. Input is a single series or
         a list of series accompanied by an optional pyarrow type to coerce the data to.
         """
-        import pyarrow as pa
-        writer = None
-        try:
-            for series in iterator:
-                batch = _create_batch(series, self._timezone, self._safecheck,
-                                      self._assign_cols_by_name)
-                if writer is None:
-                    write_int(SpecialLengths.START_ARROW_STREAM, stream)
-                    writer = pa.RecordBatchStreamWriter(stream, batch.schema)
-                writer.write_batch(batch)
-        finally:
-            if writer is not None:
-                writer.close()
+        batches = (_create_batch(series, self._timezone, self._safecheck, self._assign_cols_by_name)
+                   for series in iterator)
```

Comment:
Hm ... @BryanCutler, seems `_init_dump_stream` was added to handle `write_int(SpecialLengths.START_ARROW_STREAM, stream)` this case alone TBH. Could we just only pull out `write_int(SpecialLengths.START_ARROW_STREAM, stream)` into here and remove `_init_dump_stream` to make this logic isolated here?

It looks tricky to do it since it's all generators. I thought we could at least do something like:

```python
        batches = (_create_batch(series, self._timezone, self._safecheck, self._assign_cols_by_name)
                   for series in iterator)
        def arrow_start_written_batches():
            should_write_start_length = True
            for batch in batches:
                if should_write_start_length:
                    write_int(SpecialLengths.START_ARROW_STREAM, stream)
                    should_write_start_length = False
                yield batch
        super(ArrowStreamPandasSerializer, self).dump_stream(arrow_start_written_batches(), stream)
```

Reply:
I actually did this way first, but was worried it was more confusing. Sounds fine to me if you guys are ok with it too.

Reply:
The only issue with this is that `createDataFrame`  would need to handle `SpecialLengths.START_ARROW_STREAM` for regular and encrypted options and I'm a little hesitant to change that.  I'll keep an option in `ArrowStreamPandasSerializer` to not write the start stream for these cases.